subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -61,7 +64,7 @@ public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionCon
                 boolean deletedFiles = false;
                 boolean debugEnabled = LOGGER.isDebugEnabled();
                 for (File file : outputFileSet) {
-                    if (file.isFile()) {
+                    if (buildOutputCleanupRegistry.isOutputOwnedByBuild(file) && file.isFile()) {
```

Comment:
Let's cleaning up directories a separate PR as soon as this one is merged.
I'll change the order of `File.isFile` and `isOutputOwnedByBuild`.